### PR TITLE
Add syntax examples to  qldoc comments for AST classes

### PR DESCRIPTION
**Pull request by owen-mc**

### CODE REVIEW

File: ql/src/semmle/go/Expr.qll [hunk 1]
```diff
@@ -411,11 +546,25 @@ class TypeAssertExpr extends @typeassertexpr, Expr {
  *
  * In most cases, the subclasses `CallExpr` and `ConversionExpr` should be used
  * instead.
+ *
+ * Examples:
+ *
+ * ```go
+ * f(x)
+ * g(a, b, ...)
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:554]: Perhaps make this a real example:

```suggestion
 * g(a, b...)
```

File: ql/src/semmle/go/Expr.qll [hunk 2]
```diff
@@ -411,11 +546,25 @@ class TypeAssertExpr extends @typeassertexpr, Expr {
...
+ * string('x')
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:555]: Perhaps more realistic:

```suggestion
 * []byte("x")
```

File: ql/src/semmle/go/Expr.qll [hunk 3]
```diff
@@ -411,11 +546,25 @@ class TypeAssertExpr extends @typeassertexpr, Expr {
...
  */
 class CallOrConversionExpr extends @callorconversionexpr, Expr { }
 
 /**
  * A type conversion expression.
+ *
+ * Examples:
+ *
+ * ```go
+ * string('x')
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:566]:
```suggestion
 * []byte("x")
```

File: ql/src/semmle/go/Expr.qll [hunk 4]
```diff
@@ -1041,37 +1540,82 @@ class Name extends Expr {
   Entity getTarget() { result = target }
 }
 
-/** A simple (that is, unqualified) name. */
+/** A simple (that is, unqualified) name.
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1543]: Perhaps introduce line breaks here and below:

```suggestion
/**
 * A simple (that is, unqualified) name.
```

File: ql/src/semmle/go/Expr.qll [hunk 5]
```diff
@@ -1041,37 +1540,82 @@ class Name extends Expr {
...
+ * Examples:
+ *
+ * ```go
+ * Println
+ * ```
+ */
 class SimpleName extends Name, Ident { }
 
-/** A qualified name. */
+/** A qualified name.
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1553]:
```suggestion
/**
 * A qualified name.
```

File: ql/src/semmle/go/Expr.qll [hunk 6]
```diff
@@ -1041,37 +1540,82 @@ class Name extends Expr {
...
+ * Examples:
+ *
+ * ```go
+ * fmt.Println
+ * ```
+*/
 class QualifiedName extends Name, SelectorExpr { }
 
-/** A name referring to an imported package. */
+/** A name referring to an imported package.
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1563]:
```suggestion
/**
 * A name referring to an imported package.
```

File: ql/src/semmle/go/Expr.qll [hunk 7]
```diff
@@ -1041,37 +1540,82 @@ class Name extends Expr {
...
+ */
 class PackageName extends Name {
   override PackageEntity target;
 
   /** Gets the package this name refers to. */
   override PackageEntity getTarget() { result = target }
 }
 
-/** A name referring to a type. */
+/** A name referring to a type.
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1578]:
```suggestion
/**
 * A name referring to a type.
```

File: ql/src/semmle/go/Expr.qll [hunk 8]
```diff
@@ -1041,37 +1540,82 @@ class Name extends Expr {
...
+*/
 class QualifiedName extends Name, SelectorExpr { }
 
-/** A name referring to an imported package. */
+/** A name referring to an imported package.
+ *
+ * Examples:
+ *
+ * ```go
+ * mypackage
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1568]:
```suggestion
 * fmt
```

File: ql/src/semmle/go/Expr.qll [hunk 9]
```diff
@@ -1031,6 +1518,18 @@ class SendRecvChanTypeExpr extends @sendrcvchantypeexpr, ChanTypeExpr {
 
 /**
  * A (possibly qualified) name referring to a package, type, constant, variable, function or label.
+ *
+ * Examples:
+ *
+ * ```go
+ * Println
+ * fmt.Println
+ * mypackage
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1527]:
```suggestion
 * fmt
```

File: ql/src/semmle/go/Expr.qll [hunk 10]
```diff
@@ -1041,37 +1540,82 @@ class Name extends Expr {
...
+ */
 class TypeName extends Name {
   override TypeEntity target;
 
   /** Gets the type this name refers to. */
   override TypeEntity getTarget() { result = target }
 }
 
-/** A name referring to a value, that is, a constant, variable or function. */
+/** A name referring to a value, that is, a constant, variable or function.
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1594]:
```suggestion
/**
 * A name referring to a value, that is, a constant, variable or function.
```

File: ql/src/semmle/go/Expr.qll [hunk 11]
```diff
@@ -1041,37 +1540,82 @@ class Name extends Expr {
...
+ */
 class ValueName extends Name {
   override ValueEntity target;
 
   /** Gets the constant, variable or function this name refers to. */
   override ValueEntity getTarget() { result = target }
 }
 
-/** A name referring to a constant. */
+/** A name referring to a constant.
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1611]:
```suggestion
/**
 * A name referring to a constant.
```

File: ql/src/semmle/go/Expr.qll [hunk 12]
```diff
@@ -1092,23 +1636,44 @@ class ConstantName extends ValueName {
   }
 }
 
-/** A name referring to a variable. */
+/** A name referring to a variable.
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1639]:
```suggestion
/**
 * A name referring to a variable.
```

File: ql/src/semmle/go/Expr.qll [hunk 13]
```diff
@@ -1092,23 +1636,44 @@ class ConstantName extends ValueName {
...
+ */
 class VariableName extends ValueName {
   override Variable target;
 
   /** Gets the variable this name refers to. */
   override Variable getTarget() { result = target }
 }
 
-/** A name referring to a function. */
+/** A name referring to a function.
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1654]:
```suggestion
/**
 * A name referring to a function.
```

File: ql/src/semmle/go/Expr.qll [hunk 14]
```diff
@@ -1092,23 +1636,44 @@ class ConstantName extends ValueName {
...
+ */
 class FunctionName extends ValueName {
   override Function target;
 
   /** Gets the function this name refers to. */
   override Function getTarget() { result = target }
 }
 
-/** A name referring to a statement label. */
+/** A name referring to a statement label.
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1669]:
```suggestion
/**
 * A name referring to a statement label.
```

File: ql/src/semmle/go/Expr.qll [hunk 15]
```diff
@@ -1188,15 +1753,31 @@ private predicate isTypeExprTopDown(Expr e) {
   e = any(Ellipsis ell | isTypeExprTopDown(ell)).getOperand()
 }
 
-/** An expression referring to a type. */
+/** An expression referring to a type.
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1756]:
```suggestion
/**
 * An expression referring to a type.
```

File: ql/src/semmle/go/Expr.qll [hunk 16]
```diff
@@ -1188,15 +1753,31 @@ private predicate isTypeExprTopDown(Expr e) {
...
+ */
 class TypeExpr extends Expr {
   TypeExpr() {
     isTypeExprBottomUp(this) or
     isTypeExprTopDown(this)
   }
 }
 
-/** An expression referring to a memory location. */
+/** An expression referring to a memory location.
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1772]:
```suggestion
/**
 * An expression referring to a memory location.
```

File: ql/src/semmle/go/Expr.qll [hunk 17]
```diff
@@ -1248,7 +1829,15 @@ class ReferenceExpr extends Expr {
   }
 }
 
-/** An expression that refers to a value (as opposed to a package, a type or a statement label). */
+/** An expression that refers to a value (as opposed to a package, a type or a statement label).
```
Comment (max-schaefer) [inline ql/src/semmle/go/Expr.qll:1832]:
```suggestion
/**
 * An expression that refers to a value (as opposed to a package, a type or a statement label).
```

**Comment by owen-mc** [comment]
Thanks @max-schaefer . I've addressed your comments and added examples for Stmt.qll and Comments.qll.

File: ql/src/semmle/go/Comments.qll [hunk 18]
```diff
@@ -39,7 +62,20 @@ class CommentGroup extends @comment_group, AstNode {
 }
 
 /**
- * A program element to which a documentation comment group may be attached.
+ * A program element to which a documentation comment group may be attached,
+ * i.e. a file, a field, a specifier, a generic declaration, a function declaration
```
Comment (max-schaefer) [inline ql/src/semmle/go/Comments.qll:66]: As a general rule, we don't use "i.e." and "e.g." in doc comments.

File: ql/src/semmle/go/Comments.qll [hunk 19]
```diff
@@ -39,7 +62,20 @@ class CommentGroup extends @comment_group, AstNode {
...
+ * or a go.mod expression
+ *
+ * Examples:
```
Comment (max-schaefer) [inline ql/src/semmle/go/Comments.qll:69]: Perhaps it would make sense to include comments in the example?

Comment (max-schaefer) [reply ql/src/semmle/go/Comments.qll:69]: (Like you did below.)

**Pull request merged**